### ISSUE TITLE
Ａ question about reinitialize 

### ISSUE THREAD

**Issue by Yanxingang** [siahuat0727/RePr#9]
Hi, I feel a little confused about reinitialize.
After reinitialize, the weights of the network have changed immediately. The accuracy may go down at the reinitializing point. Could you tell me why the curve is continuous？

**Comment by siahuat0727** [comment]
Hi,
I mask the weights that connected to the pruned filters with zeros. [code](https://github.com/siahuat0727/RePr/blob/b0f46f496b0f98d082a386e1bdf71ec0a5d6f40b/main.py#L205)
Note that this operation doesn't mentioned in the paper.

**Comment by machanic** [comment]
@siahuat0727 I possibly found a bug: In the original paper, the author says there are two networks, the original network and the pruned one.  But your code seems has just one network. One consequece is that ,the prune does not work properly, because you prune the conv kernels by setting these conv kernels to 0, but after that, you train the network as usual, and the conv-layers update all their weights including the pruned one!

**Comment by machanic** [comment]
@siahuat0727 I think in order to solve this problem, you need to freeze these pruned channel inside kernel after pruning,  or create another pruned network and copy the original network's weights (except the pruned channels) to it and train this small pruned network, as the author written in that paper.

**Comment by siahuat0727** [comment]
@sharpstill 
Hi,
if I understand correctly, I already freeze it here.

https://github.com/siahuat0727/RePr/blob/b0f46f496b0f98d082a386e1bdf71ec0a5d6f40b/main.py#L68-L73

**Comment by machanic** [comment]
@siahuat0727 I still think this step has problem, because you use 
```
with torch.no_grad():
    for name, W in conv_weights:
        W.grad[mask[name]] = 0
```
but you freeze all the weights of conv-layers because `conv_weights` are all the channels' weights, according to
https://github.com/siahuat0727/RePr/blob/b0f46f496b0f98d082a386e1bdf71ec0a5d6f40b/main.py#L266
I mean you need to freeze the pruned-channels of the conv-kernel, not all of them.
I recommend you to consider my suggestion in my last response. (use a newly-created small sub-network and copy the parameters of un-pruned weight to it, then train this sub-network.)
Another question is that some conv-layers have bias, your `W.grad[mask[name]] = 0` does not consider the bias case.

**Comment by machanic** [comment]
Furthermore, the BN layer will also be affected by the channel number, and you use all the channels to train, BN layer also won't work properly.

**Comment by siahuat0727** [comment]
Hi,
I didn't freeze all of them,  `mask[name]` is the indices of pruned filters.
https://github.com/siahuat0727/RePr/blob/b0f46f496b0f98d082a386e1bdf71ec0a5d6f40b/main.py#L161-L166
I think the only difference is the bias term but I don't think that's the point. You can try for the difference.

**Comment by machanic** [comment]
Thank you for your reply, I will try my solution and report the result to compare with yours.

**Comment by siahuat0727** [comment]
Thanks. Look forward to your experiment.
